pkg/registration/hub/gc/controller.go

```diff
@@ -0,0 +1,176 @@
+	metadataClient metadata.Interface,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	clusterPatcher := patcher.NewPatcher[
+		*clusterv1.ManagedCluster, clusterv1.ManagedClusterSpec, clusterv1.ManagedClusterStatus](
+		clusterClient.ClusterV1().ManagedClusters())
+	controller := &GCController{
+		clusterLister:   clusterInformer.Lister(),
+		namespaceLister: namespaceInformer.Lister(),
+		clusterPatcher:  clusterPatcher,
+		gcReconcilers: []gcReconciler{
+			// currently only support to gc addons and works, need consider to add rbac to registration if gc other resources.
+			newGCResourcesController(metadataClient, []resourceFilter{
+				{addonGvr, ""}, {workGvr, getGCWorkLabelSelector()}}, eventRecorder),
+			newGCClusterRbacController(kubeClient, clusterPatcher, clusterInformer, clusterRoleLister,
+				clusterRoleBindingLister, roleBindingLister, manifestWorkLister, eventRecorder),
+			newGCClusterRoleController(clusterRoleLister, clusterInformer.Lister(), manifestWorkLister, kubeClient.RbacV1(), eventRecorder),
+		},
+	}
+
+	return factory.New().
+		WithInformersQueueKeysFunc(queue.QueueKeyByMetaName, clusterInformer.Informer(), namespaceInformer.Informer()).
+		WithSync(controller.sync).ToController("GCController", eventRecorder)
+}
+
+// gc controller is watching cluster and ns and to do these jobs:
+//  1. add a cleanup finalizer to managedCluster if the cluster is not deleting.
+//  2. clean up all rbac and resources in the cluster ns when the cluster is deleting
+//     or the cluster is deleted and cluster ns is deleting.
+//  3. clean up registration and work clusterRoles if there is no cluster and manifestWorks.
+func (r *GCController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	clusterName := controllerContext.QueueKey()
+	if clusterName == "" || clusterName == factory.DefaultQueueKey {
+		return nil
+	}
+
+	cluster, err := r.clusterLister.Get(clusterName)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	if cluster == nil {
+		ns, err := r.namespaceLister.Get(clusterName)
+		switch {
+		case errors.IsNotFound(err):
+			return nil
+		case err != nil:
+			return err
+		case ns.DeletionTimestamp.IsZero():
+			return nil
+		}
+	}
+
+	var errs []error
+	for index := 0; index < len(r.gcReconcilers); index++ {
+		op, err := r.gcReconcilers[index].reconcile(ctx, clusterName, cluster)
+		switch op {
+		case gcReconcileRequeue:
+			controllerContext.Queue().AddAfter(clusterName, 10*time.Second)
+			return nil
+		case gcReconcileStop:
+			return err
+		}
+
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) == 0 {
+		return nil
+	}
+
+	if cluster == nil {
+		return utilerrors.NewAggregate(errs)
+	}
+
+	newCluster := cluster.DeepCopy()
+	applyErrors := operatorhelpers.NewMultiLineAggregate(errs)
+	meta.SetStatusCondition(&newCluster.Status.Conditions, metav1.Condition{
+		Type:    ManagedClusterConditionDeleting,
+		Status:  metav1.ConditionFalse,
+		Reason:  "Error",
```

Comment:
followup no in necessarily in this PR but need a TODO: we should have more details, resourceRemaning？finalizerPending?

Reply:
added to the deleting condition